### ISSUE TITLE
Dataset.padded_batch doc improvement request

### ISSUE THREAD

**Issue by harahu** [tensorflow/tensorflow#21820]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: n/a
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: n/a
- **TensorFlow installed from (source or binary)**: n/a
- **TensorFlow version (use command below)**: 1.10
- **Python version**: 3.6
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: n/a

### Describe the problem
Documentation problem. In the method documentation for 'tf.data.Dataset.padded_batch', the argument of `padded_shapes` is described to be 

> A nested structure of `tf.TensorShape` or `tf.int64` vector tensor-like objects...

Python contains a myriad of different nesting structures. I would like the doc to at least refer to some complete definition of what this nested structure can be. Can it contain lists, tuples, dicts, sets? Maybe even more complex objects? Can it be an arbitrary non-cyclic graph? Please make clear any limitations on this structure.

### Source code / logs
[https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/data/ops/dataset_ops.py](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/data/ops/dataset_ops.py)

**Comment by tensorflowbutler** [comment]
Nagging Assignees @mrry, @MarkDaoust: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Jiri: would you mind taking over this issue that I've been shamefully neglecting?

**Comment by jsimsa** [comment]
tl;dr: I am open to concrete suggestions on how to improve the documentation.

The nesting structure of an element is derived from the structure of `tf.data` sources that support nesting (e.g. `from_tensors` and `from_tensor_slices`) and modified through `tf.data` transformations, particularly those whose output is a result of applying a user-provided function.

The supported nested constructs are `list`, `tuple`, and `dict`. Having said that, the `list` type seems to behave differently for `tf.data` sources and `tf.data` transformations. For `tf.data` sources, it results in extra tensor dimension, while for `tf.data` transformations it results in a nesting level (as if it was a `tuple`).

To illustrate this, consider the following example:
```
dataset_1 = tf.data.Dataset.from_tensors([1, 2])
dataset_2 = tf.data.Dataset.range(1).map(lambda _: [1, 2])
```
The former produces a single tensor of rank 1, while the latter produces a tuple of two tensors of rank 0.

**Comment by tgruwell** [comment]
I'd like to take a stab at this if nobody else is.

**Comment by jsimsa** [comment]
Sounds good. @aaudiber should be able to help with any questions you have as he is the current owner of tf.data API documentation.

**Comment by harahu** [comment]
> I'd like to take a stab at this if nobody else is.

I'd almost forgotten about this, but great if you do! You are the hero open source projects need.

**Comment by chunduriv** [comment]
@harahu,

The Python constructs that can be used to express the (nested) structure of elements include `tuple`, `dict`, `NamedTuple`, and `OrderedDict`. For more details you can refer [here](https://www.tensorflow.org/guide/data#dataset_structure).

**Comment by chunduriv** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)
